### PR TITLE
Feature/move to langgraph

### PR DESCRIPTION
**Pull request by jamesrichards4**
## Context

We should make use of the new Redbox class which is based on LangGraph to simplify the Redbox interface and create a library which is separate from the API to enable other uses and better experimentation

## Changes proposed in this pull request

Create a simple class Redbox which wraps the Langgraph work. Use this object in core-api in place of the current runnables.

Replace the testing in core-api with the TestCases setup from redbox-core

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [ ] I have run integration tests

### CODE REVIEW

File: core-api/core_api/routes/chat.py [hunk 1]
```diff
@@ -39,62 +37,38 @@
...
+            ],
+        ),
+    )
+    return await (redbox.graph | map_to_chat_response).ainvoke(state)
 
 
 @chat_app.get("/tools", tags=["chat"])
 async def available_tools(
-    routable_chains: Annotated[dict[str, tuple[Runnable, str]], Depends(get_routable_chains)],
+    redbox: Annotated[Redbox, Depends(get_redbox)],
```
Comment (gecBurton) [inline core-api/core_api/routes/chat.py:62]: are we using this in the front end? if not i would remove it as its more surface are to test/maintain/get-out-of-date

File: core-api/core_api/routes/chat.py [hunk 2]
```diff
@@ -39,62 +37,38 @@
...
-        ],
-    )
-
-    log.info("Routed to %s", route_name)
-    log.info("Selected files: %s", chat_request.selected_files)
-
-    return selected_chain, params
-
-
 @chat_app.post("/rag", tags=["chat"])
```
Comment (gecBurton) [inline core-api/core_api/routes/chat.py:40]: https://technologyprogramme.atlassian.net/browse/REDBOX-585

File: core-api/core_api/routes/chat.py [hunk 3]
```diff
@@ -39,62 +37,38 @@
...
 async def available_tools(
-    routable_chains: Annotated[dict[str, tuple[Runnable, str]], Depends(get_routable_chains)],
+    redbox: Annotated[Redbox, Depends(get_redbox)],
 ):
     """REST endpoint. Get a mapping of all tools available via chat."""
-    return [{"name": name, "description": tool[1]} for (name, tool) in routable_chains.items()]
+    return [{"name": name, "description": description} for name, description in redbox.get_available_keywords().items()]
 
 
 @chat_app.websocket("/rag")
```
Comment (gecBurton) [inline core-api/core_api/routes/chat.py:68]: https://technologyprogramme.atlassian.net/browse/REDBOX-585

File: redbox-core/redbox/__init__.py [hunk 4]
```diff
@@ -1,3 +1,7 @@
 """Redbox is a Python library for working with the Redbox API, data and services."""
 
+from redbox.app import Redbox
+
 __version__ = "0.3.0"
```
Comment (gecBurton) [inline redbox-core/redbox/__init__.py:5]:
```suggestion
__version__ = "0.5.0"
```

File: redbox-core/redbox/app.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    ROUTABLE_KEYWORDS = {ChatRoute.search: "Search for an answer to the question in the document"}
+
+    def __init__(
+        self,
+        llm: BaseChatModel = None,
+        all_chunks_retriever: VectorStoreRetriever = None,
+        parameterised_retriever: VectorStoreRetriever = None,
+        tokeniser: Encoding = None,
+        env: Settings = None,
+        debug: bool = False,
```
Comment (gecBurton) [inline redbox-core/redbox/app.py:38]: note-to-self: settings that define:
* llm
* all_chunks_retriever
* parameterised_retriever
* tokeniser
should be the settings that are injected into the request and not in `env = Settings`

File: redbox-core/redbox/chains/graph.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+
+@chain
+def set_route(state: ChainState):
+    """
+    Choose an approach to chatting based on the current state
+    """
+    # Match keyword
+    route_match = re_keyword_pattern.search(state["query"].question)
+    if route_match:
```
Comment (gecBurton) [inline redbox-core/redbox/chains/graph.py:32]: nitpick: walrus operator belongs here?

File: redbox-core/redbox/chains/graph.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    return chat_prompt_from_messages
+
+
+@chain
+def set_prompt_args(state: ChainState):
+    log.debug("Setting prompt args")
+    return {
+        "prompt_args": {
+            "formatted_documents": format_documents(state.get("documents") or []),
```
Comment (gecBurton) [inline redbox-core/redbox/chains/graph.py:91]:
```suggestion
            "formatted_documents": format_documents(state.get("documents", [])),
```
nitpick

File: redbox-core/redbox/graph/chat.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    app.add_edge("set_chat_prompt_args", "llm")
+
+    app.add_node("llm", build_llm_chain(llm, tokeniser, env, env.ai.chat_system_prompt, env.ai.chat_question_prompt))
+
+    return app.compile(debug=debug)
+
+
+@chain
+def set_chat_method(state: ChainState):
+    """
```
Comment (gecBurton) [inline redbox-core/redbox/graph/chat.py:67]: super clear :+1:

File: redbox-core/redbox/test/data.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            ).model_dump(),
+        )
+
+
+@dataclass
+class TestData:
+    number_of_docs: int
+    tokens_in_all_docs: int
+    chunk_resolution: ChunkResolution = ChunkResolution.largest
+    expected_llm_response: list[str] = None
```
Comment (gecBurton) [inline redbox-core/redbox/test/data.py:45]: nitpick
```suggestion
    expected_llm_response: list[str] | None = None
```

File: core-api/core_api/routes/chat.py (hunk 1 above)
Comment (jamesrichards4) [reply core-api/core_api/routes/chat.py:62]: we aren't but we should, it's in as a first step in some features I'd like to work on in the frontend

File: core-api/core_api/routes/chat.py (hunk 2 above)
Comment (jamesrichards4) [reply core-api/core_api/routes/chat.py:40]: commented on the ticket

File: redbox-core/redbox/chains/graph.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            if ChatRoute(match):
+                selected = match
+        except ValueError:
+            selected = ChatRoute.error_no_keyword.value
+    elif len(state["query"].file_uuids) > 0:
+        selected = ChatRoute.chat_with_docs.value
+    else:
+        selected = ChatRoute.chat.value
+    log.info(f"Based on user query [{selected}] selected")
+    return {"route_name": selected}
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox-core/redbox/chains/graph.py:49]: yeah any function which is a node returns a state update dictionary. In this case the state will be updated to have route_name set to whatever we selected

File: redbox-core/redbox/chains/graph.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        Create a ChatPrompTemplate as part of a chain using 'chat_history'.
+        Returns the PromptValue using values in the input_dict
+        """
+        log.debug("Setting chat prompt")
+        chat_history_budget = token_budget - len(tokeniser.encode(state["query"].question))
+
+        if chat_history_budget <= 0:
+            raise QuestionLengthError
+
+        truncated_history: list[dict[str, str]] = []
```
[the remark this answers is not in the document]
Comment (jamesrichards4) [reply redbox-core/redbox/chains/graph.py]: agreed, there's likely a way to pull this out nicely in a separate PR

File: redbox-core/redbox/chains/graph.py (hunk 7 above)
Comment (jamesrichards4) [reply redbox-core/redbox/chains/graph.py:96]: documents can be in the state dictionary as None so this function returns None, this covers both there is no documents value and documents is in there as None. There's probably a nicer way to handle state so this doesn't happen though

**Review by gecBurton — APPROVED** [review]
happy - lets fix forward

**Pull request merged**